### PR TITLE
colexec: simplify DrainMeta and Close implementations

### PR DESCRIPTION
**Pull request by asubiotto**
Since DrainMeta and Close are now called from the same goroutine as Next, the implementations can be simplified. This PR does that in a couple of commits and fixes a bug in the first commit.

Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Very nice cleanup!

> Reviewed 2 of 8 files at r1, 33 of 33 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-:-MBA49hlAOcBFt3EE765:bn0zv17)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9sJxpB9kqrXrWlMrr:-MB9sJxqA-9AY8T7wboi:b-vb6axj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/materializer.go#L64)):*
> ```Go
> 	cancelFlow func() context.CancelFunc
> 
> 	// closers is a slice of IdempotentClosers that should be Closed on
> ```

nit: `s/Idempotent//`.
___
*[pkg/sql/colexec/op_creation.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9sc-y0pfKKAXDjiud:-MB9sc-y0pfKKAXDjiue:b-c6ql93) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/op_creation.go#L81)):*
> ```Go
> 	InternalMemUsage int
> 	MetadataSources  []execinfrapb.MetadataSource
> 	// ToClose is a slice of components that need to be Closed. Close should be
> ```

nit: should the second sentence from the comment be removed?
___
*[pkg/sql/colexec/ordered_synchronizer_tmpl.go, line 245 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9zX4w08c-DSyBecic:-MB9zX4w08c-DSyBecid:b-vg9q4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colexec/ordered_synchronizer_tmpl.go#L245)):*
> ```Go
> }
> 
> func (o *OrderedSynchronizer) Close(ctx context.Context) error {
> ```

nit: we could add `var _ Closer = &OrderedSynchronizer{}` assertion.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9sujs17A3T6ykzzfn:-MB9sujs17A3T6ykzzfo:b-vb6axj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/parallel_unordered_synchronizer.go#L123)):*
> ```Go
> 	// drained in the same goroutine as Op.
> 	MetadataSources execinfrapb.MetadataSources
> 	// ToClose are IdempotentClosers in the input tree that should be closed in
> ```

nit: `s/Idempotent//`.
___
*[pkg/sql/colexec/routers.go, line 496 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t42jBLCMWKFUCNNC:-MB9t42jBLCMWKFUCNND:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L496)):*
> ```Go
> 	// drained when the HashRouter terminates.
> 	metadataSources execinfrapb.MetadataSources
> 	// closers is a slice of IdempotentClosers that need to be closed when the
> ```

ditto
___
*[pkg/sql/colexec/routers.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t9CM6KO9HJhSdYbl:-MB9t9CM6KO9HJhSdYbm:btje8lu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L713)):*
> ```Go
> 	close(r.waitForMetadata)
> 
> 	for _, closer := range r.closers {
> ```

This block appears in several places, maybe we should extract a helper method for it?
___
*[pkg/sql/colexec/serial_unordered_synchronizer.go, line 89 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9zmjn2INk_iM5NBVA:-MB9zmjn2INk_iM5NBVB:b8lgigl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colexec/serial_unordered_synchronizer.go#L89)):*
> ```Go
> }
> 
> func (s *SerialUnorderedSynchronizer) Close(ctx context.Context) error {
> ```

ditto for `_ Closer` assertion.
___
*[pkg/sql/colflow/vectorized_flow.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9tUEE5Y5oxSM7_qOG:-MB9tUEE5Y5oxSM7_qOH:b-kr7gem) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colflow/vectorized_flow.go#L124)):*
> ```Go
> 
> 	// numClosers is the number of components in the flow that implement
> 	// Close. This is used for testing assertions.
> ```

nit: maybe add a "testing assertions" or "testing state" struct for these two variables?
___
*[pkg/sql/colflow/vectorized_flow.go, line 813 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9uOab59c-Yh903uOu:-MB9uOab59c-Yh903uOv:bud7ten) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colflow/vectorized_flow.go#L813)):*
> ```Go
> // spec of pspec. The metadataSourcesQueue and toClose slices are fully consumed
> // by either passing them to an outbox or HashRouter to be drained/closed, or
> // storing it in streamIDToInputOp with the given op to be processed later.
> ```

nit: `s/storing it/storing them/`.
___
*[pkg/sql/colflow/vectorized_flow.go, line 1021 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBA-6ey1urznfgAF1gt:-MBA-6ey1urznfgAF1gu:bm2ytp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colflow/vectorized_flow.go#L1021)):*
> ```Go
> 		// TODO(asubiotto): I was looking for how to assert that Close is
> 		//  called exactly as many times as there are closers.
> 		if flowCtx.Cfg.TestingKnobs.CheckVectorizedFlowIsClosedCorrectly {
> ```

Do you think it's worth keeping this testing knob permanently?
___
*[pkg/sql/colflow/colrpc/inbox.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9vIsQ9wsaW5i5zart:-MB9vIsQ9wsaW5i5zaru:be6787c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fcba35a306ef7dd9b1554383a0e86264dda3548f/pkg/sql/colflow/colrpc/inbox.go#L90)):*
> ```Go
> 
> 	// stream is the RPC stream. It is set when RunWithStream is called but
> 	// only the Next and DrainMeta goroutines may access it.
> ```

nit: "Next and DrainMeta goroutines" are now the same, right?
___
*[pkg/sql/colflow/colrpc/inbox_test.go, line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBA3yG44Zu5LktTKhaL:-MBA3yG5BmsCrmHLZXaP:b-hn3s1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colflow/colrpc/inbox_test.go#L220)):*
> ```Go
> 		// drianMetaAfterNextIsExhausted specifies that DrainMeta should be called
> 		// after Next returns a zero-length batch.
> 		drainMetaAfterNextIsExhausted
> ```

Is it reasonable to have a scenario when `DrainMeta` is not called at all?

**Review by asubiotto — COMMENTED** [review]
Thanks! Responded to comments and fixed CI issues.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-:-MBcSSm42kQDFxryJJ0A:b-nx8pzl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/routers.go, line 496 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t42jBLCMWKFUCNNC:-MBZB8u_E355kaG-5FRX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L496)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/colexec/routers.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t9CM6KO9HJhSdYbl:-MBZedkh6XzAysDt-1p-:b1w50g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L713)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This block appears in several places, maybe we should extract a helper method for it?
</blockquote></details>

I considered this but didn't think it was worth it because 1) it's a couple of lines of code 2) extracting this would take away the ability to have callers decide what to do with the error. I can revisit this if you'd like.
___
*[pkg/sql/colexec/serial_unordered_synchronizer.go, line 89 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9zmjn2INk_iM5NBVA:-MBZeypW5e2zkCvPc1-N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colexec/serial_unordered_synchronizer.go#L89)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for `_ Closer` assertion.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 1021 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBA-6ey1urznfgAF1gt:-MBZtZ2u0kNJ8nvwvU1c:b8cjvzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colflow/vectorized_flow.go#L1021)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do you think it's worth keeping this testing knob permanently?
</blockquote></details>

I think so, what do you have in mind?
___
*[pkg/sql/colflow/colrpc/inbox_test.go, line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBA3yG44Zu5LktTKhaL:-MBZvWMz-UplIH5IE5mo:bywgkxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colflow/colrpc/inbox_test.go#L220)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is it reasonable to have a scenario when `DrainMeta` is not called at all?
</blockquote></details>

It is, although don't know when it would happen. Added.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 26 of 37 files at r5, 33 of 33 files at r6, 2 of 2 files at r7, 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-:-MBdor5M9qCBMvl7k24L:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/routers.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t9CM6KO9HJhSdYbl:-MBdnr7h6wzXR4RJs4_t:bsywoi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L713)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I considered this but didn't think it was worth it because 1) it's a couple of lines of code 2) extracting this would take away the ability to have callers decide what to do with the error. I can revisit this if you'd like.
</blockquote></details>

It seems like currently the only difference is the message when logging error, and all callers are logging the error with at the same verbosity in the same way. If in the future we do decide to have separate error handling in different callers, we could add the ability to specify that back, but right now that ability is not exercised, so, personally, I prefer a helper method in such case that is reused by all callers, but up to you.
___
*[pkg/sql/colflow/vectorized_flow.go, line 1021 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBA-6ey1urznfgAF1gt:-MBdnOdk89-UIKNDdNwp:bl2hakn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6bc1b34883aecfc4037b701aac41b2cd3b1854a/pkg/sql/colflow/vectorized_flow.go#L1021)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think so, what do you have in mind?
</blockquote></details>

I was just thinking that this knob might not be beneficial after this PR merges, but now I think it's worth keeping it. Never mind.
___
*[pkg/sql/colflow/colrpc/inbox_test.go, line 221 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MBdmHzK5cTr3nS0nziR:-MBdmHzK5cTr3nS0nziS:bovp4om) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe2951b1e50ddd1f1017a15e5762ce9b06c83520/pkg/sql/colflow/colrpc/inbox_test.go#L221)):*
> ```Go
> 		// after Next returns a zero-length batch.
> 		drainMetaAfterNextIsExhausted
> 		// drainMeta specifies that DrainMeta is not called during execution.
> ```

nit: `s/drainMeta/drainMetaNotCalled/`.

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/routers.go, line 713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50867#-MB9t9CM6KO9HJhSdYbl:-MBhY_Hy3JFSCmbFJked:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e79ce36e0b561413201bab67c1deaee0bafa02/pkg/sql/colexec/routers.go#L713)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It seems like currently the only difference is the message when logging error, and all callers are logging the error with at the same verbosity in the same way. If in the future we do decide to have separate error handling in different callers, we could add the ability to specify that back, but right now that ability is not exercised, so, personally, I prefer a helper method in such case that is reused by all callers, but up to you.
</blockquote></details>

Done.

**Comment by asubiotto** [comment]
`TestInboxShutdown` failure due to the new code addition. Will take a look and fix.

**Comment by asubiotto** [comment]
OK. Should be all good now. Will wait for green CI before borsing.

**Comment by asubiotto** [comment]
bors r=yuzefovich

**Pull request merged**
